### PR TITLE
Add attendance class

### LINKED ISSUE #120: Add Attendance Association Class
Association between `Event` and `Member`, to be contained in Member

### PR DESCRIPTION
**Pull request by andrechuakj**
Closes #120 
Created the following main classes:
* Attendance
* Hours
* Remark
* UniqueAttendanceList
* AttendanceNotFoundException
* DuplicateAttendanceException

Created the following test classes:
* AttendanceTest
* HoursTest
* RemarkTest
* UniqueAttendanceListTest
* TypicalAttendances
* AttendanceBuilder

Updated the following classes:
* CommandTestUtil: Added attendance fields
* CliSyntax: Added r/Remark and h/Hours

### CODE REVIEW

File: docs/diagrams/LogicClassDiagram.puml [hunk 1]
```diff
@@ -38,7 +38,7 @@ LogicManager --> Storage
 Storage --[hidden] Model
 Command .[hidden]up.> Storage
 Command .right.> Model
-note right of XYZCommand: XYZCommand = CreateMemberCommand, \nFindCommand, etc
+remark right of XYZCommand: XYZCommand = CreateMemberCommand, \nFindCommand, etc
```
Comment (wr1159) [inline docs/diagrams/LogicClassDiagram.puml]: Should this stay as note?
```suggestion
note right of XYZCommand: XYZCommand = CreateMemberCommand, \nFindCommand, etc
```

File: docs/diagrams/UndoRedoState5.puml [hunk 2]
```diff
@@ -18,5 +18,5 @@ State2 -[hidden]right-> State3
 class Pointer as "Current State" #FFFFFF
 
 Pointer -up-> State3
-note right on link: State ab2 deleted.
+remark right on link: State ab2 deleted.
```
Comment (wr1159) [inline docs/diagrams/UndoRedoState5.puml]: same here

File: src/main/java/seedu/address/logic/parser/AddressBookParser.java [hunk 3]
```diff
@@ -48,7 +48,7 @@ public Command parseCommand(String userInput) throws ParseException {
         final String commandWord = matcher.group("commandWord");
         final String arguments = matcher.group("arguments");
 
-        // Note to developers: Change the log level in config.json to enable lower level (i.e., FINE, FINER and lower)
+        // Remark to developers: Change the log level in config.json to enable lower level (i.e., FINE, FINER and lower)
```
Comment (wr1159) [inline src/main/java/seedu/address/logic/parser/AddressBookParser.java]: same here 
```suggestion
        // Note to developers: Change the log level in config.json to enable lower level (i.e., FINE, FINER and lower)
```

File: docs/diagrams/UndoRedoState5.puml (hunk 2 above)
Comment (wr1159) [reply docs/diagrams/UndoRedoState5.puml]:
```suggestion
note right on link: State ab2 deleted.
```

File: src/main/java/seedu/address/model/AddressBook.java [hunk 4]
```diff
@@ -24,7 +24,7 @@ public class AddressBook implements ReadOnlyAddressBook {
      * The 'unusual' code block below is a non-static initialization block, sometimes used to avoid duplication
      * between constructors. See https://docs.oracle.com/javase/tutorial/java/javaOO/initial.html
      *
-     * Note that non-static init blocks are not recommended to use. There are other ways to avoid duplication
+     * Remark that non-static init blocks are not recommended to use. There are other ways to avoid duplication
```
Comment (wr1159) [inline src/main/java/seedu/address/model/AddressBook.java]:
```suggestion
     * Note that non-static init blocks are not recommended to use. There are other ways to avoid duplication
```

File: src/main/java/seedu/address/model/attendance/Hours.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+        checkArgument(isValidHours(hours), MESSAGE_CONSTRAINTS);
+        value = Integer.parseInt(hours);
+    }
+
+    /**
+     * Returns true if a given number of hours is a valid hours.
+     */
+    public static boolean isValidHours(String test) {
+        try {
+            Integer.valueOf(test);
```
Comment (wr1159) [inline src/main/java/seedu/address/model/attendance/Hours.java]: Should we test for negative values and include it in the constrains? I believe users can enter negative numbers right now.

File: src/main/java/seedu/address/ui/EventCard.java [hunk 6]
```diff
@@ -14,7 +14,7 @@ public class EventCard extends UiPart<Region> {
     private static final String FXML = "EventListCard.fxml";
 
     /**
-     * Note: Certain keywords such as "location" and "resources" are reserved keywords in JavaFX.
+     * Remark: Certain keywords such as "location" and "resources" are reserved keywords in JavaFX.
```
Comment (wr1159) [inline src/main/java/seedu/address/ui/EventCard.java]: Remark would make sense here too, so I'll leave it up to you to decide whether to change it here

File: src/main/java/seedu/address/ui/MemberCard.java [hunk 7]
```diff
@@ -17,7 +17,7 @@ public class MemberCard extends UiPart<Region> {
     private static final String FXML = "MemberListCard.fxml";
 
     /**
-     * Note: Certain keywords such as "location" and "resources" are reserved keywords in JavaFX.
+     * Remark: Certain keywords such as "location" and "resources" are reserved keywords in JavaFX.
```
Comment (wr1159) [inline src/main/java/seedu/address/ui/MemberCard.java]: Same here

File: src/test/java/seedu/address/model/attendance/HoursTest.java [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+    @Test
+    void isValidHours() {
+
+        // invalid hours
+        assertFalse(Hours.isValidHours(null)); // null
+        assertFalse(Hours.isValidHours("")); // empty string
+        assertFalse(Hours.isValidHours(" ")); // spaces only
+        assertFalse(Hours.isValidHours("twenty")); // wrong format
+        assertFalse(Hours.isValidHours("2hrs")); // wrong format
+        assertFalse(Hours.isValidHours("3.5")); // wrong format
```
Comment (wr1159) [inline src/test/java/seedu/address/model/attendance/HoursTest.java]: I think we can add negative values here, should we add in maximum and negative values of int -  2147483647 and -2147483647, this may be overkill though but we can check for overflow then?

File: src/test/java/seedu/address/storage/StorageManagerTest.java [hunk 9]
```diff
@@ -36,7 +36,7 @@ private Path getTempFilePath(String fileName) {
     @Test
     public void prefsReadSave() throws Exception {
         /*
-         * Note: This is an integration test that verifies the StorageManager is properly wired to the
+         * Remark: This is an integration test that verifies the StorageManager is properly wired to the
```
Comment (wr1159) [inline src/test/java/seedu/address/storage/StorageManagerTest.java]: Same here with the other case of remark being valid

File: src/test/java/seedu/address/storage/StorageManagerTest.java [hunk 10]
```diff
@@ -50,7 +50,7 @@ public void prefsReadSave() throws Exception {
     @Test
     public void addressBookReadSave() throws Exception {
         /*
-         * Note: This is an integration test that verifies the StorageManager is properly wired to the
+         * Remark: This is an integration test that verifies the StorageManager is properly wired to the
```
Comment (wr1159) [inline src/test/java/seedu/address/storage/StorageManagerTest.java]: Same here

**Review by wr1159 — CHANGES_REQUESTED** [review]
LGTM Overall, left a few comments about cases in comments where note was accidentally changed to remark. Left a comment about hours seeming to accept negative values.

File: src/main/java/seedu/address/model/attendance/Attendance.java [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+            return true;
+        }
+
+        // instanceof handles nulls
+        if (!(other instanceof Attendance)) {
+            return false;
+        }
+
+        Attendance otherAttendance = (Attendance) other;
+        return memberName.equals(otherAttendance.memberName)
```
Comment (dylkaw) [inline src/main/java/seedu/address/model/attendance/Attendance.java]: Should an Attendance object be considered equal to another as long as the member and event are equal?

File: src/main/java/seedu/address/model/attendance/Hours.java (hunk 5 above)
Comment (dylkaw) [reply src/main/java/seedu/address/model/attendance/Hours.java]: I agree. We should ensure that we do not allow negative values.

File: src/test/java/seedu/address/testutil/TypicalAttendances.java [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+
+    public static final Attendance BENSON_BOXING = new AttendanceBuilder().withMemberName("Benson Meier")
+            .withEventName("Boxing Day").withHours("7").withRemark("Bring boxing gloves").build();
+
+    public static final Attendance CARL_CNY = new AttendanceBuilder().withMemberName("Carl Kurz")
+            .withEventName("Chinese New Year").withHours("3").withRemark("Brought red packets").build();
+    public static final Attendance DANIEL_DOG_DAY = new AttendanceBuilder().withMemberName("Daniel Meier")
+            .withEventName("Dog Celebration Day").withHours("5").withRemark("Bringing dogs").build();
+
+    public static final Attendance ELLE_EHCO_DAY = new AttendanceBuilder().withMemberName("Elle Meyer")
```
Comment (dylkaw) [inline src/test/java/seedu/address/testutil/TypicalAttendances.java]:
```suggestion
    public static final Attendance ELLE_ECHO_DAY = new AttendanceBuilder().withMemberName("Elle Meyer")
```
Just a minor typo here.

**Review by dylkaw — CHANGES_REQUESTED** [review]
Other than a comment of the equals method of Attendance, LGTM!

File: docs/diagrams/LogicClassDiagram.puml (hunk 1 above)
Comment (andrechuakj) [reply docs/diagrams/LogicClassDiagram.puml]: Yup it should

File: src/main/java/seedu/address/model/attendance/Attendance.java (hunk 11 above)
Comment (wr1159) [reply src/main/java/seedu/address/model/attendance/Attendance.java]: That's a good point, I agree with that

File: src/main/java/seedu/address/model/attendance/Hours.java (hunk 5 above)
Comment (andrechuakj) [reply src/main/java/seedu/address/model/attendance/Hours.java]: Good suggestion, I will add that

File: src/test/java/seedu/address/model/attendance/HoursTest.java (hunk 8 above)
Comment (andrechuakj) [reply src/test/java/seedu/address/model/attendance/HoursTest.java]: Yup I'll add in testing of overflow. Do you think we should include that in our message constraint? (E.g. "Hours should be a non-negative integer up to 2147483647 and it should not be blank")

File: src/main/java/seedu/address/model/attendance/Attendance.java (hunk 11 above)
Comment (andrechuakj) [reply src/main/java/seedu/address/model/attendance/Attendance.java]: Agree, I will add in the method and update the test cases

File: src/test/java/seedu/address/model/attendance/HoursTest.java (hunk 8 above)
Comment (wr1159) [reply src/test/java/seedu/address/model/attendance/HoursTest.java]: Hmm maybe we don't add it in the message constrain but the UG? I think it won't look really nice if we display a large number, but let's get the others' opinion on this.

**Review by dylkaw — APPROVED** [review]

**Review by wr1159 — APPROVED** [review]
LGTM! Thanks
